allensdk/brain_observatory/ecephys/ecephys_project_api/http_engine.py

```diff
@@ -1,23 +1,192 @@
+        for ii, chunk in enumerate(response.iter_content(self.chunksize)):
             if ii == 0:
                 size_message = f"{response_mb:3.3}mb" if response_mb is not None else "potentially large"
                 logging.warning(f"downloading a {size_message} file from {url}")
-            yield chunk
\ No newline at end of file
+            yield chunk
+
+            elapsed = time.perf_counter() - start_time
+            if elapsed > self.timeout:
+                raise requests.Timeout(f"Download took {elapsed} seconds, but timeout was set to {self.timeout}")
+
+
+class AsyncHttpEngine(HttpEngine):
+
+    def __init__(
+        self, 
+        scheme: str, 
+        host: str, 
+        session: Optional[aiohttp.ClientSession] = None, 
+        **kwargs
+    ):
+        """ Simple tool for making asynchronous streaming http requests.
+
+        Parameters
+        ----------
+        scheme :
+            e.g "http" or "https"
+        host : 
+            will be used as the base for request urls
+        session : 
+            If provided, this preconstructed session will be used rather than 
+            a new one. Keep in mind that AsyncHttpEngine closes its session 
+            when it is garbage collected!
+        **kwargs :
+            Will be passed to parent.
+
+        """
+
+        super(AsyncHttpEngine, self).__init__(scheme, host, **kwargs)
+
+        if session:
+            self.session = session
+            warnings.warn(
+                "Recieved preconstructed session, ignoring timeout parameter."
+            )
+        else:
+            self.session = aiohttp.ClientSession(
+                timeout=aiohttp.client.ClientTimeout(self.timeout)
+            )
+
+    async def _stream_coroutine(self, route, callback):
+        url = self._build_url(route)
+
+        async with self.session.get(url) as response:
+            await callback(response.content.iter_chunked(self.chunksize))
+
+    def stream(self, route):
+        """ Returns a coroutine which
+            - makes an http request
+            - exposes internally an asynchronous iterator over the response
+            - takes a callback parameter, which should consume the iterator.
+
+        Parameters
+        ----------
+        route :
+            the http route (under this object's host) to request against.
+
+        """
+
+        return functools.partial(self._stream_coroutine, route)
+
+    def __del__(self):
+        if hasattr(self, "session"):
+            nest_asyncio.apply()
+            loop = asyncio.get_event_loop()
+            loop.run_until_complete(self.session.close())
+        
+
+def write_bytes_from_coroutine(path: str, coroutine):
+    """ Utility for streaming http from an asynchronous requester to a file.
+
+    Parameters
+    ----------
+    path : 
+        Write to this file
+    coroutine : 
+        Ought to accept a callback which itself accepts an asynchronous 
+        iterator.
+    
```

Comment:
I think for this method to be in the public API we should have a clearer docstring. It requires a specific coroutine that in turn requires a specific callback. I would include a code example in this docstring that uses the AsyncHTTPEngine, and clarify the requirements of the coroutine and the callback it requires.

Reply:
I think it is really helpful when your methods return functions to include examples in the docstring (see `functools.partial`)
Add something like this:
```
Example:
>>> session = AsyncHttpEngine(scheme="http", host="myhost")
>>> write_bytes_from_coroutine(path="path/to/file.json", session.stream("myroute"))
```
Because this also is essentially serving as the documentation for how to actually use AsyncHTTPEngine, since there is no instance method to easily write data from a stream.